source/agora/node/Ledger.d

```diff
@@ -98,54 +106,154 @@ public class Ledger
         Called when a new transaction is received.
 
         If the transaction is accepted it will be added to
-        a new block, and the block will be added to the ledger.
+        the transaction pool. If there are enough valid transactions
+        in the pool, a block will be created.
 
         If the transaction is invalid, it's rejected and false is returned.
 
         Params:
             tx = the received transaction
 
         Returns:
-            true if the transaction is valid and was added to a block
+            true if the transaction is valid and was added to the pool
 
     ***************************************************************************/
 
     public bool acceptTransaction (Transaction tx) @safe
     {
-        if (!tx.isValid(&this.findUTXO))
+        if (!this.isValidTransaction(tx))
+        {
+            logInfo("Rejected tx: %s", tx);
             return false;
+        }
 
         this.pool.add(tx);
         if (this.pool.length >= Block.TxsInBlock)
-            this.makeBlock();
+            this.tryCreateBlock();
 
         return true;
     }
 
     /***************************************************************************
 
-        Create a new block out of transactions in the storage.
+        Add a validated block to the Ledger,
+        and add all of its outputs to the UTXO set.
+
+        Params:
+            block = the block to add
+
+    ***************************************************************************/
+
+    private void addValidatedBlock (const ref Block block) nothrow @safe
+    {
+        // add the new UTXOs
+        block.txs.each!(tx => this.utxo_set.updateUtxoCache(tx));
+
+        // remove the TXs from the Pool
+        block.txs.each!(
+            (tx)
+            {
+                try
+                {
+                    this.pool.remove(tx.hashFull());
+                }
+                catch (Exception ex)
+                {
+                    // this OP must not fail
+                    assert(0);
+                }
+            }
+        );
+
+        if (!this.storage.saveBlock(block))
+            assert(0);
+
+        // read back and cache the last block
+        if (!this.storage.readLastBlock(this.last_block))
+            assert(0);
+    }
+
+    /***************************************************************************
+
+        Try making a new block if there are enough valid and non double-spending
+        transactions in the pool
+
+        Double-spending transactions will be skipped over while iterating
+        over the pool. If there are not enough valid transactions,
+        a block will not be created.
```

Comment:
Hum, should double spend end up in the pool ? We have an issue for that, maybe link it.

Reply:
Actually this is a really good point, because it opens a DoS vector.